### DISCUSSION TITLE
Regression in RGroupDecomposition

### DISCUSSION THREAD

**Discussion by bp-kelley** [rdkit/rdkit#7896] [Development]
@ptosco @greglandrum 

Commit https://github.com/rdkit/rdkit/commit/6c966b4fe5b1d6b23b53cbc2356cba55ac06dad5 seems to really slow down the rgroup decomposition.  I have a test case (I'll see if I can get it to you all) with 10k compounds and a benzene core.

Running it before this commit:

```
len(mols)=10000
#    time  #accepted
100 0.23 18
200 0.43 -1
300 1.17 45
400 2.11 -1
500 2.70 -1
600 3.33 -1
700 5.15 -1
800 5.45 -1
```

And after

```
#    time  #accepted
100 1.97 18
200 3.92 -1
300 11.90 45
400 21.77 -1
500 28.02 -1
600 34.93 -1
700 54.53 -1
800 57.65 -1
```

Note I'm 99% certain it was this commit, but will double check.  I expect it is some data structure change causing the exponential behavior that should be easy to pick up on a profiler, but I may pull you in if I can't sort it out.

**Comment by ptosco** [comment]
@bp-kelley @greglandrum I have done a quick internal test on a demanding internal dataset and I can reproduce the slowdown.
The change that causes the slowdown is the removal of the following `if` clause (I have slightly cleaned it down since `data->permutationProduct` is not used anywhere except in this `if` clause):
```
  if (data->params.matchingStrategy != GA) {
    size_t N = 1;
    for (auto matche = data->matches.begin() + data->previousMatchSize;
         matche != data->matches.end(); ++matche) {
      size_t sz = matche->size();
      N *= sz;
    }
    // oops, exponential is a pain
    if (N * potentialMatches.size() > 100000) {
      data->process(data->prunePermutations);
    }
  }
```
It is not surprising that in the absence of the `if` clause the code becomes slower as the number of loaded molecules increases: in fact, as far as I can understand, the `if` clause introduces a change in behavior from `GreedyChunks` to `Greedy` if the Cartesian product size exceed 10^5.
We can certainly restore the previous behavior, which however seems very questionable to me, as it is based on an arbitrary threshold. I wonder whether we should rethink the algorithm rather than just reverting to the old behavior because it is faster. I would rather prefer to use `Greedy` from the beginning than having the algorithm change behavior arbitrarily if it deems the calculation to be too onerous.

**Reply by bp-kelley** [reply]
There is going to be an arbitrary nature here to some extent.  The threshold was arbitrary but only in the sense of performance, the comment was poorly worded, but basically we needed to choose something so that run time didn't explode.

One approach that I was playing with but got a bit lost, is that when N molecules have already been processed, we explicitly reduce the chunk size (immediately or slowly) to 1 since comparing the current batch of molecules to itself outside of the already processed ones won't make any difference.  This is essentially what this cutoff is doing but would be in a more controlled manner.

All of these will be heuristics, and relay on the  assumption that if we already have a lot of data processed, we don't have to work as hard with newer data.  In anycase, the performance regression is currently unfavorable so I recommend reverting this particular change as it will impact production codes (and already has).

**Reply by bp-kelley** [reply]
btw - thanks for identifying the root cause, I'm sorry I didn't catch it in review.

**Comment by ptosco** [comment]
I have submitted https://github.com/rdkit/rdkit/pull/7898 to restore that code, minus a variable that was being updated without being referenced anywhere. However, I don't think that restoring that code, which introduces an undocumented change in behavior, and is not exercised in any of the tests, is the best way to alleviate performance issues in the RGD algorithm.

**Comment by bp-kelley** [comment]
Removing the code caused an undocumented change in behavior, this is just restoring it.  We can discuss a proper solution in the future.
